lib/libutee/tee_api_arith_mpi.c

```diff
@@ -0,0 +1,837 @@
+
+static void __noreturn mpi_panic(const char *func, int line, int rc)
+{
+	printf("Panic function %s, line %d, code %d\n", func, line, rc);
+	TEE_Panic(0xB16127 /*BIGINT*/);
+	while (1)
+		; /* Panic will crash the thread */
+}
+
+#define MPI_CHECK(x) do { \
+		int _rc = (x); \
+		 \
+		if (_rc) \
+			mpi_panic(__func__, __LINE__, _rc); \
+	} while (0)
+
+void _TEE_MathAPI_Init(void)
+{
+	static uint8_t data[MPI_MEMPOOL_SIZE]
+		__aligned(__alignof__(mbedtls_mpi_uint));
+
+	mbedtls_mpi_mempool = mempool_alloc_pool(data, sizeof(data), NULL);
+	if (!mbedtls_mpi_mempool)
+		API_PANIC("Failed to initialize memory pool");
+}
+
+struct bigint_hdr {
+	int32_t sign;
+	uint16_t alloc_size;
+	uint16_t nblimbs;
+};
+
+static void init_static_mpi(mbedtls_mpi *mpi, TEE_BigInt *bigInt)
+{
+	struct bigint_hdr *hdr = (struct bigint_hdr *)bigInt;
+
+	mbedtls_mpi_init_static(mpi, (mbedtls_mpi_uint *)(hdr + 1),
+				hdr->sign, hdr->alloc_size, hdr->nblimbs);
+}
+
+/*
+ * Initializes a MPI.
+ *
+ * If a bigint is supplied it's initialized with the value of the bigint
+ * and changes will be written back completely with a call to put_mpi().
+ * The bigint dictates the size of the MPI which will be fixed to this
+ * size.
+ *
+ * If no bigint is supplied a temporary MPI is allocated instead which will
+ * be freed by put_mpi().
+ */
+static void get_mpi(mbedtls_mpi *mpi, TEE_BigInt *bigInt)
+{
+	/*
+	 * The way the GP spec is defining the bignums it's
+	 * difficult/tricky to do it using 64-bit arithmetics given that
+	 * we'd need 64-bit alignment of the data as well.
+	 */
+	COMPILE_TIME_ASSERT(sizeof(mbedtls_mpi_uint) == sizeof(uint32_t));
+
+	if (bigInt)
+		init_static_mpi(mpi, bigInt);
+	else
+		mbedtls_mpi_init_mempool(mpi);
+}
+
+/*
+ * Initializes a MPI from a constant bigint.
+ *
+ * A MPI is allocated and given an initial value based on the supplied
+ * bigint. When the MPI is freed with put_mpi() no changes are propagated
+ * back.
+ */
+static void get_const_mpi(mbedtls_mpi *mpi, const TEE_BigInt *bigInt)
+{
+	mbedtls_mpi mpi_const;
+
+	init_static_mpi(&mpi_const, (TEE_BigInt *)bigInt);
+	get_mpi(mpi, NULL);
+	MPI_CHECK(mbedtls_mpi_copy(mpi, &mpi_const));
+}
+
+/*
+ * Uninitialize a MPI.
+ *
+ * If the MPI is linked to a bigint the final changes (size and sign) will
+ * be copied back.
+ *
+ * If the MPI isn't linked to bigint it's only freed.
+ */
+static void put_mpi(mbedtls_mpi *mpi)
+{
+	if (mpi->alloc_type == MBEDTLS_MPI_ALLOC_TYPE_STATIC) {
+		struct bigint_hdr *hdr = ((struct bigint_hdr *)mpi->p) - 1;
+
+		hdr->sign = mpi->s;
+		hdr->nblimbs = mpi->n;
+	} else {
+		mbedtls_mpi_free(mpi);
+	}
+}
+
+void TEE_BigIntInit(TEE_BigInt *bigInt, uint32_t len)
+{
+	memset(bigInt, 0, TEE_BigIntSizeInU32(len) * sizeof(uint32_t));
+
+	struct bigint_hdr *hdr = (struct bigint_hdr *)bigInt;
+
+	hdr->sign = 1;
+	if ((len - 2) > MBEDTLS_MPI_MAX_LIMBS)
+		API_PANIC("Too large bigint");
+	hdr->alloc_size = len - 2;
```

Comment:
Could you  replace this `2` with some macro that reflects something like `(ROUNDUP(sizeof(struct bigint_hdr), sizeof(uint32_t)) / sizeof(uint32_t))`.

This comments also applies to the entry of functions `bigint_binary_mod()` , `TEE_BigIntMod()` and `TEE_BigIntInvMod()`.

Reply:
That's a very complicated way of writing 2, this value is a requirement by GP.
I'll add a BIGINT_HDR_SIZE_IN_U32 define and a compile time assert to check that it's matching with the header.

The `2` in bigint_binary_mod() , TEE_BigIntMod() and TEE_BigIntInvMod() is used in comparison with the value of a bigint and is a GP requirement.

Reply:
ok, thanks for the explanation and sorry for false positive.